From 89070cb31f9e0513943024f172f58f01cd8167c2 Mon Sep 17 00:00:00 2001
From: "GONG, Ruiqi" <gongruiqi1@huawei.com>
Date: Thu, 10 Nov 2022 17:39:37 +0800
Subject: [PATCH 0017/1398] [Huawei] ARM: spectre-v2: add config isolation for
 this feature

Offering: HULK
hulk inclusion
category: feature
bugzilla: 180851 https://gitee.com/openeuler/kernel/issues/I4EF1O

----------------------------------------

Add CONFIG_ARM_NOSPECTRE_V2 to separate code of this feature from
baseline.

Signed-off-by: GONG, Ruiqi <gongruiqi1@huawei.com>
---
 arch/arm/mm/Kconfig        | 7 +++++++
 arch/arm/mm/proc-v7-bugs.c | 8 ++++++--
 arch/arm/mm/proc-v7.S      | 8 ++++++++
 3 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mm/Kconfig b/arch/arm/mm/Kconfig
index c6bf34a33849..951e33c43373 100644
--- a/arch/arm/mm/Kconfig
+++ b/arch/arm/mm/Kconfig
@@ -1141,3 +1141,10 @@ config DEBUG_ALIGN_RODATA
 	  additional section-aligned split of rodata from kernel text so it
 	  can be made explicitly non-executable. This padding may waste memory
 	  space to gain the additional protection.
+
+config ARM_NOSPECTRE_V2
+	bool "`nospectre_v2=` cmdline support for 32-bit ARM"
+	default n
+	help
+	  This option allows ARM32 systems to turn off Spectre mitigation via
+	  `nospectre_v2=` kernel cmdline for better performance.
diff --git a/arch/arm/mm/proc-v7-bugs.c b/arch/arm/mm/proc-v7-bugs.c
index 4f25d230d2f6..2858bd056a17 100644
--- a/arch/arm/mm/proc-v7-bugs.c
+++ b/arch/arm/mm/proc-v7-bugs.c
@@ -9,20 +9,22 @@
 #include <asm/spectre.h>
 #include <asm/system_misc.h>
 
+#ifdef CONFIG_ARM_NOSPECTRE_V2
 /*
  * 32-bit ARM spectre hardening, enabled by default, can be disabled via boot
  * cmdline param 'nospectre_v2' to avoid performance regression.
  */
 int nospectre_v2 __read_mostly;
 
-static int nospectre_bhb __read_mostly;
-
 static int __init nospectre_v2_setup(char *str)
 {
 	nospectre_v2 = 1;
 	return 0;
 }
 early_param("nospectre_v2", nospectre_v2_setup);
+#endif
+
+static int nospectre_bhb __read_mostly;
 
 static int __init nospectre_bhb_setup(char *str)
 {
@@ -140,10 +142,12 @@ static void cpu_v7_spectre_v2_init(void)
 {
 	unsigned int state, method = 0;
 
+#ifdef CONFIG_ARM_NOSPECTRE_V2
 	if (nospectre_v2) {
 		pr_info_once("Spectre v2: hardening is disabled\n");
 		return;
 	}
+#endif
 
 	switch (read_cpuid_part()) {
 	case ARM_CPU_PART_CORTEX_A8:
diff --git a/arch/arm/mm/proc-v7.S b/arch/arm/mm/proc-v7.S
index 2fcffcc60cc6..291b31d575d3 100644
--- a/arch/arm/mm/proc-v7.S
+++ b/arch/arm/mm/proc-v7.S
@@ -113,21 +113,29 @@ ENDPROC(cpu_v7_hvc_switch_mm)
 #endif
 
 ENTRY(cpu_v7_iciallu_switch_mm)
+#ifdef CONFIG_ARM_NOSPECTRE_V2
 	ldr_l r3, nospectre_v2
 	cmp r3, #1
 	beq 1f
+#endif
 	mov	r3, #0
 	mcr	p15, 0, r3, c7, c5, 0		@ ICIALLU
+#ifdef CONFIG_ARM_NOSPECTRE_V2
 1:
+#endif
 	b	cpu_v7_switch_mm
 ENDPROC(cpu_v7_iciallu_switch_mm)
 ENTRY(cpu_v7_bpiall_switch_mm)
+#ifdef CONFIG_ARM_NOSPECTRE_V2
 	ldr_l r3, nospectre_v2
 	cmp r3, #1
 	beq 1f
+#endif
 	mov	r3, #0
 	mcr	p15, 0, r3, c7, c5, 6		@ flush BTAC/BTB
+#ifdef CONFIG_ARM_NOSPECTRE_V2
 1:
+#endif
 	b	cpu_v7_switch_mm
 ENDPROC(cpu_v7_bpiall_switch_mm)
 
-- 
2.17.1

